opnode/rollup/sync/start.go

```diff
@@ -59,45 +60,94 @@ var TooDeepReorgErr = errors.New("reorg is too deep")
 
 const MaxReorgDepth = 500
 
-// FindSafeL2Head takes the supplied L2 start block and walks the L2 chain until it finds the first L2 block reachable from the supplied
-// block that is also canonical.
-func FindSafeL2Head(ctx context.Context, start eth.BlockID, l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (eth.L2BlockRef, error) {
-	// Starting point
-	l2Head, err := l2.L2BlockRefByHash(ctx, start.Hash)
-	if err != nil {
-		return eth.L2BlockRef{}, fmt.Errorf("failed to fetch L2 head: %w", err)
+// isCanonical returns true if the supplied block ID is canonical in the L1 chain.
+// It will suppress ethereum.NotFound errors
+func isCanonical(ctx context.Context, l1 L1Chain, block eth.BlockID) (bool, error) {
+	canonical, err := l1.L1BlockRefByNumber(ctx, block.Number)
+	if err != nil && !errors.Is(err, ethereum.NotFound) {
+		return false, err
+	} else if err != nil {
+		return false, nil
 	}
+	return canonical.Hash == block.Hash, nil
+}
+
+// FindL2Heads walks back from the supplied L2 blocks and finds the unsafe and safe L2 heads.
+// Unsafe Head: The first L2 block who's L1 Origin is canonical (determined via block by number).
+// Safe Head: Walk back 1 sequence window of epochs from the Unsafe Head.
+func FindL2Heads(ctx context.Context, start eth.L2BlockRef, seqWindowSize int,
+	l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (unsafeHead eth.L2BlockRef, safeHead eth.L2BlockRef, err error) {
 	reorgDepth := 0
-	// Walk L2 chain from L2 head to first L2 block which has a L1 Parent that is canonical. May walk to L2 genesis
-	for n := l2Head; ; {
-		l1header, err := l1.L1BlockRefByNumber(ctx, n.L1Origin.Number)
-		if err != nil {
-			// Generic error, bail out.
-			if !errors.Is(err, ethereum.NotFound) {
-				return eth.L2BlockRef{}, fmt.Errorf("failed to fetch L1 block %v: %w", n.L1Origin.Number, err)
-			}
-			// L1 block not found, keep walking chain
-		} else {
-			// L1 Block found, check if matches & should keep walking the chain
-			if l1header.Hash == n.L1Origin.Hash {
-				return n, nil
+	var prevL1OriginHash common.Hash
+	// First check if the L1 Origin of the start block is ahead of the current L1 head
+	// If so, we assume that this should be the next unsafe head for the sequencing window
+	// We still need to walk back the safe head because we don't know where the reorg started.
+	l1Head, err := l1.L1HeadBlockRef(ctx)
+	if err != nil {
+		return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+	}
+	l2Ahead := start.L1Origin.Number > l1Head.Number
+
+	// Walk L2 chain from start until L1Origin matches l1Base. Bail out early of when at genesis.
+	for n := start; ; {
+		// Check if l1Origin is canonical when we get to a new epoch
+		if prevL1OriginHash != n.L1Origin.Hash {
+			if ok, err := isCanonical(ctx, l1, n.L1Origin); err != nil {
+				return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+			} else if ok {
+				unsafeHead = n
+				break
```

Comment:
The `unsafeHead` should only change when the L1 information of that block is *not* canonical. It should stay the same as long as the L1 info is not-found. Changing it here when the L1 info is canonical could result in unnecessary reorgs.

Reply:
This is why I don't like named returns in go. See line 130 for how this returns correctly if the start is ahead of L1